### PR TITLE
Add support for MODULARIZE with USE_PTHREADS

### PR DESCRIPTION
**Pull request by juj**
This adds MODULARIZE support with USE_PTHREADS. Will probably need some iteration, putting this up already to get a CI run on this, and since this was asked for by some people before.

### CODE REVIEW

File: src/settings.js [hunk 1]
```diff
@@ -829,6 +829,12 @@ var MODULARIZE = 0;
 // (since you arean't creating the instance yourself).
 var MODULARIZE_INSTANCE = 0;
 
+// If we separate out asm.js with the --separate-asm option,
+// this is the name of the variable where the generated asm.js
+// Module is assigned to. This name can either be a property
+// of Module, or a freestanding variable name, like "var asmJsModule".
+var ASM_MODULE_NAME = 'Module["asm"]';
+
```
Comment (kripken) [inline src/settings.js]: this part could be a separate PR, I think?

File: tools/shared.py [hunk 2]
```diff
@@ -3057,13 +3057,52 @@ def safe_copy(src, dst):
     return
   shutil.copyfile(src, dst)
 
+def find_lines_with_preprocessor_directives(file):
+  return filter(lambda x: '#if' in x or '#elif' in x, open(file, 'r').readlines())
+
+def run_c_preprocessor_on_file(src, dst):
```
Comment (kripken) [inline tools/shared.py]: I think this clang preprocessor change seems unnecessary in this PR? (Also, I was hoping we don't need this after the recent preprocessor eval improvement?)

File: src/shell.js [hunk 3]
```diff
@@ -81,11 +81,26 @@ if (Module['ENVIRONMENT']) {
...
+  var STATICTOP = {{{EXPORT_NAME}}}.STATICTOP;
+  var DYNAMIC_BASE = {{{EXPORT_NAME}}}.DYNAMIC_BASE;
+  var DYNAMICTOP_PTR = {{{EXPORT_NAME}}}.DYNAMICTOP_PTR;
+  var PthreadWorkerInit = {{{EXPORT_NAME}}}.PthreadWorkerInit;
+  var STACK_BASE = {{{EXPORT_NAME}}}.STACK_BASE;
+  var STACKTOP = {{{EXPORT_NAME}}}.STACKTOP;
+  var STACK_MAX = {{{EXPORT_NAME}}}.STACK_MAX;
+}
+
+var currentScriptUrl = typeof _scriptDir !== 'undefined' ? _scriptDir : ((typeof document !== 'undefined' && document.currentScript) ? document.currentScript.src : undefined);
```
[the remark this answers is not in the document]
Comment (juj) [reply src/shell.js]: This is pulling in the `document.currentScript` variable from very top of the script evaluation time. Found a better way to do this that retains one reference only to `document.currentScript`, much better that way.

File: src/library_pthread.js [hunk 4]
```diff
@@ -1177,3 +1178,8 @@ var LibraryPThread = {
 
 autoAddDeps(LibraryPThread, '$PThread');
 mergeInto(LibraryManager.library, LibraryPThread);
+
+if (USE_PTHREADS) {
+  DEFAULT_LIBRARY_FUNCS_TO_INCLUDE.push('emscripten_futex_wake');
```
[the remark this answers is not in the document]
Comment (juj) [reply src/library_pthread.js]: That is a good question - when I was developing Unity to work against this, I needed to add these in, but checking test cases now, I cannot reproduce this, so dropped these until I run into a scenario again that shows that these would be needed.

File: src/postamble.js [hunk 5]
```diff
@@ -464,6 +464,9 @@ if (!ENVIRONMENT_IS_PTHREAD) // EXIT_RUNTIME=0 only applies to default behavior
 
 #if USE_PTHREADS
 if (!ENVIRONMENT_IS_PTHREAD) run();
+
+Module._emscripten_futex_wake = _emscripten_futex_wake;
```
[the remark this answers is not in the document]
Comment (juj) [reply src/postamble.js]: (same as above, dropped)

File: src/settings.js (hunk 1 above)
Comment (juj) [reply src/settings.js]: #7949

File: tools/shared.py (hunk 2 above)
Comment (juj) [reply tools/shared.py]: Migrated to use the newer preprocessor evaling code

**Comment by juj** [comment]
> I worry about the EmscriptenCode change here - that seems like a breaking change?

Unless I did a bug, the `EmscriptenCode` change should apply only if a) user is building to .html and b) user did not pass -s EXPORT_NAME directive at all. My impression is that such a build mode would generate a broken build anyways, since before this change, it would generate a page that does not work.

If user builds to .js or explicitly passes in `-s EXPORT_NAME=` directive of their own, the change should not apply.

The main reason I did that is that with such change, I could edit `src/settings.js` and flip `MODULARIZE = 0;` to `MODULARIZE = 1;` and be able to run `python tests/runner.py browser.test_pthread*` to mass test all pthread tests in modularize enabled mode.

Though I can revert that if it's not desirable, I don't need it for anything anymore now that I have run the tests.

> Overall, I wonder if a way to reduce complexity here might be to say that pthreads **requires** modularize?

I don't like that change, if we did that, I think we should require overall that all builds require MODULARIZE, otherwise we are diverging. I think it is ok for now.

File: emcc.py [hunk 6]
```diff
@@ -1952,8 +1971,8 @@ def repl(m):
 
       if shared.Settings.USE_PTHREADS:
         target_dir = os.path.dirname(os.path.abspath(target))
-        shutil.copyfile(shared.path_from_root('src', 'worker.js'),
-                        os.path.join(target_dir, shared.Settings.PTHREAD_WORKER_FILE))
+        worker_output = os.path.join(target_dir, shared.Settings.PTHREAD_WORKER_FILE)
+        open(worker_output, 'w').write(shared.read_and_preprocess(shared.path_from_root('src', 'worker.js'), expand_macros=True))
```
[the remark this answers is not in the document]
Comment (juj) [reply emcc.py]: ok

**Comment by brion** [comment]
Couple of things I noticed testing this branch with ogv.js:

If I compile with `-s ERROR_ON_UNDEFINED_SYMBOLS=1` I get `error: undefined symbol: emscripten_webgl_create_context` when building my modules. (Don't have a minimal test case handy yet, can write one up later.) Easily worked around by setting it to 0 for now.

I get `Uncaught TypeError: Cannot set property '_stdin' of undefined` when instantiating a module for a second time, here:

```
    var _stdin;
        if (ENVIRONMENT_IS_PTHREAD)
            _stdin = PthreadWorkerInit._stdin;
        else
            PthreadWorkerInit._stdin = _stdin = 40384;
```

Also, a variable ENVIRONMENT_IS_PTHREAD leaks into the window global.... that may be causing this bit to fail:

```
if (typeof ENVIRONMENT_IS_PTHREAD === "undefined") {
            ENVIRONMENT_IS_PTHREAD = false;
            var PthreadWorkerInit = {}
        } else {
            var buffer = OGVDecoderVideoVP9MTW.buffer;
            var tempDoublePtr = OGVDecoderVideoVP9MTW.tempDoublePtr;
            var TOTAL_MEMORY = OGVDecoderVideoVP9MTW.TOTAL_MEMORY;
            var STATICTOP = OGVDecoderVideoVP9MTW.STATICTOP;
            var DYNAMIC_BASE = OGVDecoderVideoVP9MTW.DYNAMIC_BASE;
            var DYNAMICTOP_PTR = OGVDecoderVideoVP9MTW.DYNAMICTOP_PTR;
            var PthreadWorkerInit = OGVDecoderVideoVP9MTW.PthreadWorkerInit
        }
```

so it's taking the second path and initializing `PthreadWorkerInit` to what turns out to be `undefined` instead of an empty object.

**Comment by brion** [comment]
Great, I'll whip up tests cases tonight or tomorrow for the multiple instantiation. :)

My code's highly module dependent so I can't test the non module path much myself.

**Comment by juj** [comment]
CI passes on this (modulo the unrelated test_debuginfo), so I hope this could land soon @kripken ? This can further be revised in tree, and this PR carries the support for preprocessing src/worker.js, which my other PRs need, so I'm a bit blocked from authoring further.

File: emcc.py [hunk 7]
```diff
@@ -1081,6 +1081,15 @@ def check(input_file):
...
       shared.Settings.MODULARIZE = 1
 
+    if shared.Settings.MODULARIZE:
+      assert not options.proxy_to_worker, '-s MODULARIZE=1 and -s MODULARIZE_INSTANCE=1 are not compatible with --proxy-to-worker (if you want to run in a worker with -s MODULARIZE=1, you likely want to do the worker side setup manually)'
+
+    if not shared.Settings.EXPORT_NAME:
+      # If we are building directly to .html with -s MODULARIZE=1 (but not -s MODULARIZE_INSTANCE=1), default to exporting
+      # Emscripten code under function EmscriptenCode() so that it does not collide with the Module object. (When not modularizing,
+      # or if building with MODULARIZE_INSTANCE, or if building to .js manually, there is no conflict, so default to 'Module' in those cases)
+      shared.Settings.EXPORT_NAME = 'EmscriptenCode' if final_suffix == '.html' and shared.Settings.MODULARIZE and not shared.Settings.MODULARIZE_INSTANCE else 'Module'
```
Comment (kripken) [inline emcc.py]: This is a breaking change - current users doing `Module()` would no longer work.

How about requiring EXPORT_NAME be set, when MODULARIZE?

Comment (juj) [reply emcc.py]: That works as well - let me update to that.

File: src/settings.js [hunk 8]
```diff
@@ -936,7 +936,10 @@ var EXPLICIT_ZEXT = 0;
 
 // Global variable to export the module as for environments without a
 // standardized module loading system (e.g. the browser and SM shell).
-var EXPORT_NAME = 'Module';
+// Default EXPORT_NAME is 'Module', but if building directly to .html
+// with -s MODULARIZE=1 but not with -s MODULARIZE_INSTANCE=1, then default EXPORT_NAME
+// is 'EmscriptenCode' to not conflict with the default provided Module object.
+var EXPORT_NAME = '';
```
Comment (kripken) [inline src/settings.js]: the comment text is not fully clear, and I still don't get it myself maybe - so we have the MODULARIZE function, and the Module object. What do we use the Module object for, that conflicts here? In MODULARIZE I would hope that it is used for nothing, and any existing uses are bugs we should fix? I know you fixed the separate-asm one, are there more?

File: emcc.py [hunk 9]
```diff
@@ -2942,18 +2962,28 @@ def un_src(self):
     """Use this if you want to modify the script and need it to be inline."""
     if self.src is None:
       return
+    run_module = ''
+    if shared.Settings.MODULARIZE and not shared.Settings.MODULARIZE_INSTANCE:
+      run_module = 'script.onload = function() { %s(Module); };' % shared.Settings.EXPORT_NAME
```
[the remark this answers is not in the document]
Comment (juj) [reply emcc.py]: If I build with
```
emcc tests\hello_world.c -s MODULARIZE=1 -o a.html -s EXPORT_NAME=MyModule
```
what should happen? Do we expect to generate a web page that does not run at all? (but user has to open up web console and run `MyModule(Module);` there? That may be a bit odd out of the box behavior. The above change makes it so that a `-o a.html` page compiled with -s MODULARIZE=1 will generate an output that runs automatically.

If developer is targeting a `-o a.js` build, then the above code is avoided, and they can have their "I'll run it manually" feature.

I suppose I am a bit indifferent about this, I can revert this if this may be problematic - just thought I'd make these kind of `-s MODULARIZE=1 -o a.html` builds run out of the box by default.

File: src/parseTools.js [hunk 10]
```diff
@@ -1477,6 +1477,34 @@ function makeStaticString(string) {
   return '(stringToUTF8("' + string + '", ' + ptr + ', ' + len + '), ' + ptr + ')';
 }
 
+// Generates access to module exports variable in pthreads worker.js
+function makeAsmExportAccessInPthread(variable) {
+  if (MODULARIZE) {
+    return "Module['" + variable + "']" // 'Module' is defined in worker.js local scope, so not EXPORT_NAME in this case.
```
[the remark this answers is not in the document]
Comment (juj) [reply src/parseTools.js]: hmm, `{{{ EXPORT_NAME }}}` is the name of the function that one needs to call to instantiate the compiled JS module - I should not be assigning anything over it?

`src/worker.js` currently does

```js
Module = {{{ EXPORT_NAME }}}(Module);
```
i.e. it uses `Module` as both the import object name, and also as export object name, like non-MODULARIZE-non-pthreads builds do.

File: src/settings.js (hunk 8 above)
Comment (juj) [reply src/settings.js]: `src/shell.html` and `src/shell_minimal.html` both have

```js
var Module = {
  ...
};
```

and we have tons of code patterns showing `Module` to be a global configuration object that has both the imports and exports. So this puts out an idiom that people will write `var Module = { ... };` configuration objects in their own shells as well.

So then if I do `-s MODULARIZE=1 -o a.html`, I will get `a.js` have `function Module(Module) {`, but my `a.html` also has `var Module = { configs here };`, which contradict each other.

Or if I do `-s MODULARIZE=1 -o a.js` and wrote my own `a.html` to launch it, following tacit practices shown by `src/shell.html` that showcases that one should write a `var Module = { ... };` config object, then again these two conflict, as my `a.html` has a `var Module = { ... };`, but when the compiled `a.js` file is loaded in, it will override it with a `function Module(...) {...}`.

The above aspects can lead to confusion of "what on earth is 'Module'?". Sometimes it is a global config object that has both imports+exports+library_x.js specific config vars. Sometimes it is the name of the JS module function that is generated (as EXPORT_NAME='Module' was default value in -s MODULARIZE=1 mode). In MODULARIZE=1 mode inside the generated module it is again the imports object (i.e. `function {{{EXPORT_NAME}}}(Module) { ... }`). In `MODULARIZE_INSTANCE=1` builds it is first the imports object, but then is replaced by assigning the export object over it.

Requiring the value of EXPORT_NAME= to come directly from the user and never giving it a default value (especially `EXPORT_NAME='Module'`) is a good step towards clearing the confusion, since people will then less easily mix these together.

File: emcc.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (kripken) [reply emcc.py]: > Do we expect to generate a web page that does not run at all?

Yes, exactly. Only a function `MyModule` should exist in the global scope, and the user is expected to call it in a way that makes sense for their application. Consider that some user may only wany to  call `MyModule` after some special input event, for example. And some users want to call it more than once.

The other behavior, where the code is run, is what `MODULARIZE_INSTANCE` does: it automatically runs it for you (and exactly once). It sounds like you're expecting `MODULARIZE` to work the same, but the use cases are different.

File: src/parseTools.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (kripken) [reply src/parseTools.js]: > hmm, {{{ EXPORT_NAME }}} is the name of the function that one needs to call to instantiate the compiled JS module - I should not be assigning anything over it?

True, yes - but I thought we were inside that function, so creating an alias might have simplified things? Did I get that wrong?

That is, what I meant to suggest is that for internal uses, we can do

```
function ExportName(Module) {
  var ExportName = Module;
  [.. internal uses here can use either Module or ExportName ..]
}
```
I think we do something similar in the general case, so that ExportName can be used internally.

File: src/settings.js (hunk 8 above)
Comment (kripken) [reply src/settings.js]: I see, very good points.

I think we should fix this before this PR. What I'd recommend is requiring `EXPORT_NAME` to not be `Module`. This would be a breaking change, so we need to prepare users, but it definitely seems necessary given the above - otherwise we'll be stuck trying to work around this in complicated ways in our codebase forever. (Sad we didn't realize this earlier, but better now than later.)

Specifically, what I'm suggesting is that the default value of `EXPORT_NAME` be something like `EmscriptenCode` as you suggested. (or maybe just `Code`? probably too short) and in emcc.py we assert on it not being set to `Module`, with a clear error.

If you don't have time for this, I could try to find time myself.

**Comment by juj** [comment]
I changed this PR to avoid the issue altogether, and only concentrate on the MODULARIZE+USE_PTHREADS combination. Adapted the tests to use a custom pattern to run. How does this look now?

File: emcc.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (juj) [reply emcc.py]: MODULARIZE_INSTANCE is different than MODULARIZE in that MODULARIZE_INSTANCE assigns export over to `{{{ EXPORT_NAME }}}` and kills the module function, whereas MODULARIZE returns the exports, and the function can be called many times. I was not expecting MODULARIZE to work like MODULARIZE_INSTANCE, but I thought that `-o a.html` would be good to be able to run out of the box, while retaining the module-like structure.

File: src/parseTools.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (juj) [reply src/parseTools.js]: The worker global scope in `src/worker.js` is not inside the function scope, but it is in the worker's global scope outside the function, so it will either need to assign variables to global scope to import to the main JS code, and read variables off of global scope or Module, if not building with MODULARIZE. If building with MODULARIZE, imports need to go in an imports object, and exports need to be read off the return value of invoking the module function. I'm not sure how applicable the code snippet you write would be here?

File: src/settings.js (hunk 8 above)
Comment (juj) [reply src/settings.js]: Reverted away from doing any changes to this, and just focused on the MODULARIZE+USE_PTHREADS combination.

**Comment by juj** [comment]
Looks green now on CI, ready for second review?

File: src/parseTools.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (kripken) [reply src/parseTools.js]: Ok, so if I understand you now, src/worker.js adds code to the global scope in both modularize and non-modularize mode?

File: emcc.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (kripken) [reply emcc.py]: Maybe I still don't get your point here, but "run out of the box" is exactly what modularize is supposed to **not** do? Making it run automatically would be a breaking and surprising change. I don't think that's different for HTML - the user would still be writing code to call the function themselves, so if you call it that would interfere.

File: src/parseTools.js [hunk 11]
```diff
@@ -1477,6 +1477,34 @@ function makeStaticString(string) {
...
+  } else {
+    return EXPORT_NAME + "['" + variable + "']"
+  }
+}
+
+// Generates access to a global scope variable in pthreads worker.js
+function makeAsmGlobalAccessInPthread(variable) {
+  if (MODULARIZE) {
+    return "Module['" + variable + "']" // 'Module' is defined in worker.js local scope, so not EXPORT_NAME in this case.
```
Comment (kripken) [inline src/parseTools.js]: missing `;` on these returns

File: src/parseTools.js [hunk 12]
```diff
@@ -1477,6 +1477,34 @@ function makeStaticString(string) {
   return '(stringToUTF8("' + string + '", ' + ptr + ', ' + len + '), ' + ptr + ')';
 }
 
+// Generates access to module exports variable in pthreads worker.js
```
Comment (kripken) [inline src/parseTools.js]: these comments could be better - should explain the stuff we are discussing here. in particular, why pthreads is different.

File: src/parseTools.js [hunk 13]
```diff
@@ -1477,6 +1477,34 @@ function makeStaticString(string) {
...
+  if (MODULARIZE) {
+    return "Module['" + variable + "']" // 'Module' is defined in worker.js local scope, so not EXPORT_NAME in this case.
+  } else {
+    return variable
+  }
+}
+
+// Generates access to both global scope variable and exported Module variable, e.g. "Module['foo'] = foo" or just plain "foo" depending on if we are MODULARIZEing.
+// Used the be able to initialize both variables at the same time.
+function makeAsmExportAndGlobalAccessInPthread(variable) {
```
Comment (kripken) [inline src/parseTools.js]: This is for an assignment target, I think? The name could specify that, and so it would not be odd that one code path has side effects and one seems to not (as in all cases this is the assign target, so there are side effects).

How about `makeAsmExportAndGlobalAssignTargetInPthread`?

File: src/shell.js [hunk 14]
```diff
@@ -81,11 +81,34 @@ if (Module['ENVIRONMENT']) {
 // 2) We could be the application main() thread proxied to worker. (with Emscripten -s PROXY_TO_WORKER=1) (ENVIRONMENT_IS_WORKER == true, ENVIRONMENT_IS_PTHREAD == false)
 // 3) We could be an application pthread running in a worker. (ENVIRONMENT_IS_WORKER == true and ENVIRONMENT_IS_PTHREAD == true)
 #if USE_PTHREADS
```
Comment (kripken) [inline src/shell.js]: please move the pthreads code from shell into a separate file - I think that will help a lot with readability here

File: src/parseTools.js (hunk 11 above)
Comment (juj) [reply src/parseTools.js]: These generate generic expressions to be used for assingment or read, so it's not a `return Module['variable'];` that would be generated here.

File: src/parseTools.js (hunk 13 above)
Comment (juj) [reply src/parseTools.js]: ok

File: src/parseTools.js (hunk 12 above)
Comment (juj) [reply src/parseTools.js]: updated the comments with more detailed info

File: src/shell.js (hunk 14 above)
Comment (juj) [reply src/shell.js]: ok

File: src/parseTools.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (juj) [reply src/parseTools.js]: Yeah, the worker does add code to global scope in both cases. From inside the JS local scope in the module in MODULARIZE mode, it reads variables in the global scope. Strictly speaking this is not polluting, since the global scope of a pthread-hosting worker is not "public", i.e. there's no main .html file or external JS code that would potentially get injected to it. But perhaps in the future it may make sense to either formalize this, or move items to the imports object. I don't know yet.

File: emcc.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (juj) [reply emcc.py]: Gotcha - yeah, then it sounds like the right option is to abandon the autorun when building to html, so this PR no longer has that.

File: src/parseTools.js (hunk 11 above)
Comment (kripken) [reply src/parseTools.js]: I mean, outside of the string:
```
return "Module['" + variable + "']";
```

Comment (juj) [reply src/parseTools.js]: ohh, gotcha :)

**Comment by VirtualTim** [comment]
So I went to try this out, but noticed that it's not part of 1.38.27. Is this planned for 1.38.28? Any ETA on when this should land?

**Comment by juj** [comment]
This should be part of 1.38.27. Git log looks like this:

**Comment by VirtualTim** [comment]
Hm, OK, it looks like me emcc.py just didn't update on my machine, even though `emsdk list` shows `emscripten-1.38.27` and `sdk-1.38.27-64bit` as installed and activated.

I had a look at the tags, and [emcc.py from Feb 6 was tagged as 1.38.27](https://github.com/emscripten-core/emscripten/commit/63810da86ec14373a8b45c5c2d0bb82e3501f645?diff=unified), and these changes came in a day later. That made me think that these changes were not included. Manually downloading the 1.38.27 zip from github showed the expected emcc.py.

So I'm not sure what's going on on my machine, apologies for the false report.
